### ISSUE TITLE
improve DB cleanup runtime

### ISSUE THREAD

**Issue by ztNFny** [cgeo/cgeo#12204]
Some quick feedback: after removing "DataStore.cleanIfNeeded(this);" from "init()" startup is very fast, can't notice any lags elsewhere. So at least for me this is the main culprit. #11227 didn't have any noticable impact, both with and without DB cleaning.

I can take a look at optimizing the SQL query, but it'll always take some time. As that functions deletes caches after 3 days I think running it only once a day or even less often would be an easy to implement workaround that doesn't have many downsides.

_Originally posted by @ztNFny in https://github.com/cgeo/cgeo/issues/11227#issuecomment-980641566_

**Comment by ztNFny** [comment]
Let's see whether this SQLite query can be improved:

                    queryToColl(dbTableCaches,
                        new String[]{"geocode"},
                        "detailedupdate < ? AND visiteddate < ? AND geocode NOT IN (SELECT DISTINCT (geocode) FROM " + dbTableCachesLists + ")",
                        new String[]{timestampString, timestampString},
                        null,
                        null,
                        geocodes,
                        GET_STRING_0);

**Comment by eddiemuc** [comment]
There's already an index on the cache table on (visiteddate, detailedupdate), so the obvious performance increasements are already there.

@ztNFny just for my interest, would you try if turning on WAL would resolve the problem. You just have to enter one line of code in method `DataStore.initAndCheck(final boolean force)` as depicted in the following screenshot.

database.enableWriteAheadLogging();


![image](https://user-images.githubusercontent.com/6909759/144044891-9ae331c8-f7d6-47a9-96ad-e4482f2d8da8.png)

**Comment by ztNFny** [comment]
Now I have the problem that after re-enabling "DataStore.cleanIfNeeded(this);" it's always fast... So I kinda broke my scenario, trying to figure out where and how 😧

**Comment by Lineflyer** [comment]
Question:
Today I have received two support mails from users who are on our nightly builds. Both claim, that starting c:geo is slow (>1 minuten) starting with 2021.11.30-NB while it was OK with 2021.11.28-NB. This is very unusual we we normally don't receive any regular feedback from nightly users.

Any related change in `master` last days?

**Comment by Lineflyer** [comment]
Additional info:
One user additionally mentioned that also using Menu -> Report a problem takes ages to complete.

**Comment by ztNFny** [comment]
> Any related change in master last days?

Only changes I see are @eddiemuc logging stuff, but that shouldn't affect performance...

https://github.com/cgeo/cgeo/commit/cf67ba02d5698abc674cbba81666ee3cbf09026f
https://github.com/cgeo/cgeo/commit/e2bb388fc615bae45bbd52d901ec143f80fb4180

**Comment by eddiemuc** [comment]
> Additional info:
> One user additionally mentioned that also using Menu -> Report a problem takes ages to complete.
> 

One of my recent changes adds some db info to system information. While I have no clue how this should impact performance notably, I don't want to exclude anything...

User Logs would be great, preferably after debug mode is activated

**Comment by ztNFny** [comment]
just pulled master into my branch, can't confirm either reported problem

**Comment by ztNFny** [comment]
while looking into the favorite / owner info missing bug I had a very slow start, unfortunately without debugging enabled:

11-30 17:57:25.727 23036 23036 I cgeo.geocachin: The ClassLoaderContext is a special shared library.
11-30 17:57:25.734 23036 23036 I cgeo.geocachin: failed to find target package for overlay /system/product/overlay/ProductOverlay-oneplus-framework-res.apk
11-30 17:57:25.734 23036 23036 I cgeo.geocachin: failed to find target package for overlay /system/system_ext/overlay/CommonOverlay-oneplus-framework-res.apk
11-30 17:57:25.734 23036 23036 I cgeo.geocachin: failed to find target package for overlay /odm/overlay/OdmOverlay-oneplus-framework-res.apk
11-30 17:57:25.744 23036 23036 I cgeo    : [main] [CtxLog]ContentStorage.init:START
11-30 17:57:25.747 23036 23036 I cgeo    : [main] [CtxLog]ContentStorage.init:END (3ms)
11-30 17:57:25.750 23036 23036 I cgeo    : [Log] Logging set: minLevel=WARN, minAddCallerInfo=NONE, addCallerInfoMaxDepth=8, throwOnError=false
11-30 17:57:26.066 23036 23036 I cgeo    : [Log] Logging set: minLevel=WARN, minAddCallerInfo=NONE, addCallerInfoMaxDepth=8, throwOnError=false
11-30 17:57:26.066 23036 23036 I cgeo    : [Log] No logging config file 'log-properties.txt' found at /cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles], using defaults
11-30 17:57:26.066 23036 23036 I cgeo    : [main] ---------------- CGeoApplication: startup -------------
11-30 17:57:26.066 23036 23036 I cgeo    : [main] [CtxLog]CGeoApplication.onCreate:START
11-30 17:57:26.074 23036 23036 W cgeo.geocachin: Accessing hidden field Landroid/view/ViewConfiguration;->sHasPermanentMenuKey:Z (greylist-max-p, reflection, denied)
11-30 17:57:26.077 23036 23036 I cgeo    : [main] [CtxLog]DataStore.init:START
11-30 17:57:26.081 23036 23036 I cgeo    : [main] [DB] Current Database Version: 98
11-30 17:57:35.720  1485  5564 E ANDR-RAMBOOST_SERVER: ramboost cmd_len = 26 cmd: iop_record cgeo.geocaching
11-30 17:57:35.720  1485  5564 I ANDR-RAMBOOST_SERVER: iop_record cgeo.geocaching 1
11-30 17:57:41.798 23036 23036 I cgeo    : [main] [DB] DB initialized: Path: /data/user/0/cgeo.geocaching/databases/data, V:98, Size:471.8 MB on system internal storage, WAL-enabled: false, IntegrityOk: true, ReadOnly: false, Page Size: 4096, Max Size: 4.0 TB
11-30 17:57:41.798 23036 23036 I cgeo    : [main] [CtxLog]DataStore.init:END (15721ms)
11-30 17:57:41.809 23036 23036 I cgeo    : [main] LooperLogger: started for Thread: main
11-30 17:57:41.809 23036 23036 I cgeo    : [main] [CtxLog]CGeoApplication.onCreate:END (15743ms)

**Comment by eddiemuc** [comment]
Seems relatively clear that creating the new db info takes unacceptable long time. 15 seconds in above example. Don't understand why.

 I will remove it again.

**Comment by eddiemuc** [comment]
Update: created PR #12207 for removal of logging additional db info

**Comment by ztNFny** [comment]
> Checked the system info of those reporting users. Look at the DB size....wow..

that's not that large ... My "full" db is 900M, my normal one 450M

**Comment by Lineflyer** [comment]
Oh..I have 1400 caches stored at 22M, how many caches is your 900M?

**Comment by ztNFny** [comment]
about 50k

**Comment by DaTechniker** [comment]
that may fit.
it`s 1,1 GB with around 70700.

**Comment by SammysHP** [comment]
It's always interesting to see how other people use c:geo. My average number of stored caches is ~20. Only caches planned for the current trip and once they are logged online at the end of the day they get deleted.

**Comment by ztNFny** [comment]
>  My average number of stored caches is ~20

That'll only get me through the first hour of the day ;-)

My lists are: Found + Homezone + Solved Mysteries + interesting tours + statistically relevant caches (e.g. the next DT matrixes, old caches, many FPs, high wilson rating, ...) + special tools (climbing, boat, ...)

**Comment by eddiemuc** [comment]
> Question: Today I have received two support mails from users who are on our nightly builds. Both claim, that starting c:geo is slow (>1 minuten) starting with 2021.11.30-NB while it was OK with 2021.11.28-NB. This is very unusual as we normally don't receive any regular feedback from nightly users.
> 
> Any related change in `master` last days?

Can anyone confirm whether PR #12207 fixed this (should be in last nights nightly)?

**Comment by ztNFny** [comment]
> Can anyone confirm whether PR #12207 fixed this (should be in last nights nightly)?

I had pulled that yesterday and it worked for me (but startup wasn't always but only sometimes slow for me)

**Comment by Lineflyer** [comment]
> Can anyone confirm whether PR #12207 fixed this (should be in last nights nightly)?

Yes, one user just confirmed, that todays nightly is starting fast again for him.

**Comment by eddiemuc** [comment]
> > Can anyone confirm whether PR #12207 fixed this (should be in last nights nightly)?
> 
> Yes, one user just confirmed, that todays nightly is starting fast again for him.

Good to hear. Strange that requesting such trivial information from database object takes that much time though...
